domain/src/main/kotlin/com/krzykrucz/elesson/currentlesson/domain/attendance/AttendanceWorkflowImplementation.kt

```diff
@@ -0,0 +1,76 @@
+package com.krzykrucz.elesson.currentlesson.domain.attendance
+
+import arrow.core.Either
+import arrow.core.left
+import arrow.core.right
+import com.krzykrucz.elesson.currentlesson.domain.startlesson.ClassRegistry
+import com.krzykrucz.elesson.currentlesson.domain.startlesson.LessonIdentifier
+import java.time.LocalDateTime
+
+fun noteAbsence(
+        isLessonStarted: IsLessonStarted,
+        isInRegistry: IsInRegistry,
+        areAllStudentsChecked: AreAllStudentsChecked
+): NoteAbsence = { lessonId, uncheckedStudent, attendanceList, classRegistry ->
+    noteStudent(isLessonStarted, lessonId, isInRegistry, uncheckedStudent, classRegistry, attendanceList, areAllStudentsChecked, addAbsentStudent())
+}
+
+fun notePresence(
+        isLessonStarted: IsLessonStarted,
+        isInRegistry: IsInRegistry,
+        areAllStudentsChecked: AreAllStudentsChecked
+): NotePresence = { lessonId, uncheckedStudent, attendanceList, classRegistry ->
+    noteStudent(isLessonStarted, lessonId, isInRegistry, uncheckedStudent, classRegistry, attendanceList, areAllStudentsChecked, addPresentStudent())
+}
+
+fun noteLate(
+        isNotTooLate: IsNotTooLate
+): NoteLate = { lessonId, absentStudent, completedAttendance ->
+    if (isNotTooLate(lessonId.lessonStartTime, LocalDateTime.now())) {
+        val attendance = completedAttendance.attendance
+        val updatedAbsentStudents = attendance.absentStudents.minusElement(absentStudent)
+        val updatedLateStudents = attendance.lateStudents.plusElement(absentStudent.toLate())
+        val updatedAttendance = attendance.copy(
+                absentStudents = updatedAbsentStudents,
+                lateStudents = updatedLateStudents
+        )
+        completedAttendance.copy(
+                attendance = updatedAttendance
+        )
+    } else {
+        completedAttendance
+    }
+}
+
+private fun <T: Student> noteStudent(
+        isLessonStarted: IsLessonStarted,
+        lessonId: LessonIdentifier,
+        isInRegistry: IsInRegistry,
+        student: T,
+        classRegistry: ClassRegistry,
+        attendanceList: AttendanceList,
+        areAllStudentsChecked: AreAllStudentsChecked,
+        updateAttendanceList: (AttendanceList, T) -> AttendanceList
+): Either<AttendanceError, Attendance> =
```

Comment:
creating such a generic function overcomplicates things, it's really hard to trace only a marking present flow or marking late flow

Reply:
Yeah, I had it at first but it seemed like a lot of repetition. After all, I agree, would revert that change.